ClaRa: Undeclared identifier at C compilation

### Description

When compiling `ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace` there are a bunch of undelcated identifiers in the C code, see error log [ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace.err](https://libraries.openmodelica.org/branches/master/ClaRa_dev/files/ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace.err).

```
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:690:47: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp8, ((modelica_real*)&(mills1._coalIn._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                              ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1287:48: error: use of undeclared identifier 'coalFlowSource'
  real_array_create(&tmp14, ((modelica_real*)&(coalFlowSource._fuelObject._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1312:48: error: use of undeclared identifier 'coalFlowSource'
  real_array_create(&tmp20, ((modelica_real*)&(coalFlowSource._fuelObject._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1337:48: error: use of undeclared identifier 'coalFlowSource'
  real_array_create(&tmp26, ((modelica_real*)&(coalFlowSource._fuelObject._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1362:48: error: use of undeclared identifier 'coalFlowSource'
  real_array_create(&tmp32, ((modelica_real*)&(coalFlowSource._fuelObject._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1389:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp38, ((modelica_real*)&(mills1._coalIn._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1414:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp44, ((modelica_real*)&(mills1._coalIn._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1439:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp50, ((modelica_real*)&(mills1._coalIn._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1464:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp56, ((modelica_real*)&(mills1._coalIn._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1736:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp62, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1761:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp68, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1799:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp74, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1824:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp80, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1849:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp86, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1874:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp92, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1899:48: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp98, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                               ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1924:49: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp104, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                                ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1949:49: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp110, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
                                                ^
ClaRa_dev_ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace_06inz.c:1974:49: error: use of undeclared identifier 'mills1'
  real_array_create(&tmp116, ((modelica_real*)&(mills1._coalOut._fuelModel._xi_e_waf)), 2, (_index_t)0, (_index_t)0);
```

### Steps to Reproduce

Load TILMedia and ClaRa deverloper version and build `combineMillandFurnace`.
```
loadFile("/path/to/TILMediaClaRa/TILMedia/package.mo"); getErrorString();
loadFile("/path/to/clara_om/ClaRa/package.mo"); getErrorString();
buildModel(ClaRa.Components.Mills.HardCoalMills.Check.combineMillandFurnace); getErrorString();
```

### Expected Behavior

Compile the model.

### Version and OS

<!--- Please complete the following information. -->

 - OpenModelica Version: v1.20.0-dev-103-ge4e4270543
 - OS: Ubuntu Focal
 - Versions of used Modelica libraries: ClaRa dev @ d931ea39e5a56704911d2d4fb78d4386e13ec553

--- a/OMCompiler/Compiler/FrontEnd/Expression.mo
+++ b/OMCompiler/Compiler/FrontEnd/Expression.mo
@@ -11636,6 +11636,25 @@ algorithm
   end matchcontinue;
 end dimensionsList;
 
+public function hasZeroDimension
+  "Returns true if at least one dimension is zero or dimensions is empty list."
+  input DAE.Dimensions inDims;
+  output Boolean hasZeroDimension = false;
+protected
+  list<Integer> intDims;
+algorithm
+  if listLength(inDims) == 0 then
+    hasZeroDimension := true;
+    return;
+  end if;
+  intDims := dimensionsList(inDims);
+  for dim in intDims loop
+    if dim == 0 then
+      hasZeroDimension := true;
+      break;
+    end if;
+  end for;
+end hasZeroDimension;
 
 public function expDimensionsList
   "Extracts a list of integers from a list of expressions"

--- a/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
+++ b/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
@@ -5253,7 +5253,7 @@ template daeExpCrefRhsSimContext(Exp ecr, Context context, Text &preExp,
         '((modelica_<%type%>*)&(<%nosubname%>))'
       let t = '<%type%>_array_create(&<%wrapperArray%>, <%arrayData%>, <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
       let &preExp += t
-      wrapperArray
+    wrapperArray
     else
       let &sub = buffer ""
       let dimsLenStr = listLength(crefDims(cr))

--- a/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
+++ b/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
@@ -5243,13 +5243,21 @@ template daeExpCrefRhsSimContext(Exp ecr, Context context, Text &preExp,
     let arrayType = type + "_array"
     let wrapperArray = tempDecl(arrayType, &varDecls)
     if crefSubIsScalar(cr) then
-      let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
-      let dimsLenStr = listLength(dims)
-      let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
-      let nosubname = contextCref(crefStripSubs(cr), context, &preExp, &varDecls, &auxFunction, &sub)
-      let t = '<%type%>_array_create(&<%wrapperArray%>, ((modelica_<%type%>*)&(<%nosubname%>)), <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
-      let &preExp += t
-    wrapperArray
+      if hasZeroDimension(dims) then
+        let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
+        let dimsLenStr = listLength(dims)
+        let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
+        let t = '<%type%>_array_create(&<%wrapperArray%>, NULL, <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
+        let &preExp += t
+      wrapperArray
+      else
+        let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
+        let dimsLenStr = listLength(dims)
+        let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
+        let nosubname = contextCref(crefStripSubs(cr), context, &preExp, &varDecls, &auxFunction, &sub)
+        let t = '<%type%>_array_create(&<%wrapperArray%>, ((modelica_<%type%>*)&(<%nosubname%>)), <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
+        let &preExp += t
+      wrapperArray
     else
       let &sub = buffer ""
       let dimsLenStr = listLength(crefDims(cr))

--- a/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
+++ b/OMCompiler/Compiler/Template/CodegenCFunctions.tpl
@@ -5243,20 +5243,16 @@ template daeExpCrefRhsSimContext(Exp ecr, Context context, Text &preExp,
     let arrayType = type + "_array"
     let wrapperArray = tempDecl(arrayType, &varDecls)
     if crefSubIsScalar(cr) then
-      if hasZeroDimension(dims) then
-        let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
-        let dimsLenStr = listLength(dims)
-        let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
-        let t = '<%type%>_array_create(&<%wrapperArray%>, NULL, <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
-        let &preExp += t
-      wrapperArray
+      let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
+      let dimsLenStr = listLength(dims)
+      let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
+      let arrayData = if hasZeroDimension(dims) then
+        'NULL'
       else
-        let &sub = buffer '<%indexSubs(crefDims(cr), crefSubs(crefArrayGetFirstCref(cr)), context, &preExp, &varDecls, &auxFunction)%>'
-        let dimsLenStr = listLength(dims)
-        let dimsValuesStr = (dims |> dim => '(_index_t)<%dimension(dim, context, &preExp, &varDecls, &auxFunction)%>' ;separator=", ")
         let nosubname = contextCref(crefStripSubs(cr), context, &preExp, &varDecls, &auxFunction, &sub)
-        let t = '<%type%>_array_create(&<%wrapperArray%>, ((modelica_<%type%>*)&(<%nosubname%>)), <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
-        let &preExp += t
+        '((modelica_<%type%>*)&(<%nosubname%>))'
+      let t = '<%type%>_array_create(&<%wrapperArray%>, <%arrayData%>, <%dimsLenStr%>, <%dimsValuesStr%>);<%\n%>'
+      let &preExp += t
       wrapperArray
     else
       let &sub = buffer ""

--- a/OMCompiler/Compiler/Template/SimCodeTV.mo
+++ b/OMCompiler/Compiler/Template/SimCodeTV.mo
@@ -3785,11 +3785,15 @@ package Expression
     output list<Integer> outValues;
   end dimensionsList;
 
+  function hasZeroDimension
+    input DAE.Dimensions inDims;
+    output Boolean hasZeroDimension;
+  end hasZeroDimension;
+
   function expDimensionsList
     input list<DAE.Exp> inDims;
     output list<Integer> outValues;
   end expDimensionsList;
-
 
   function isMetaArray
     input DAE.Exp inExp;

